pkg/resources/management.cattle.io/v3/projectroletemplatebinding/validator.go

```diff
@@ -152,50 +142,65 @@ func clusterFromProject(project string) (string, string) {
-		invalidFieldName = "groupName"
+		return field.Invalid(fieldPath.Child("groupName"), newPRTB.GroupName, reason)
 	case oldPRTB.GroupPrincipalName != newPRTB.GroupPrincipalName && oldPRTB.GroupPrincipalName != "":
-		invalidFieldName = "groupPrincipalName"
+		return field.Invalid(fieldPath.Child("groupPrincipalName"), newPRTB.GroupPrincipalName, reason)
 	case (newPRTB.GroupName != "" || oldPRTB.GroupPrincipalName != "") && (newPRTB.UserName != "" || oldPRTB.UserPrincipalName != ""):
-		invalidFieldName = "both user and group"
+		return field.Forbidden(fieldPath,
+			"binding must target either a user [userName]/[userPrincipalName] OR a group [groupName]/[groupPrincipalName]")
+	case newPRTB.ServiceAccount != "":
+		return field.Forbidden(fieldPath.Child("serviceAccount"), "update is not allowed")
 	default:
 		return nil
 	}
-
-	return fmt.Errorf("cannot update %s for clusterRoleTemplateBinding %s: %w", invalidFieldName, oldPRTB.Name, admission.ErrInvalidRequest)
 }
 
 // validateCreateFields checks if all required fields are present and valid.
-func (a *admitter) validateCreateFields(newPRTB *apisv3.ProjectRoleTemplateBinding) error {
+func (a *admitter) validateCreateFields(newPRTB *apisv3.ProjectRoleTemplateBinding, fieldPath *field.Path) error {
+	if newPRTB.ServiceAccount != "" {
+		return field.Forbidden(fieldPath.Child("serviceAccount"), "")
+	}
+
 	hasUserTarget := newPRTB.UserName != "" || newPRTB.UserPrincipalName != ""
 	hasGroupTarget := newPRTB.GroupName != "" || newPRTB.GroupPrincipalName != ""
 
 	if (hasUserTarget && hasGroupTarget) || (!hasUserTarget && !hasGroupTarget) {
-		return fmt.Errorf("binding must target either a user [userId]/[userPrincipalId] OR a group [groupId]/[groupPrincipalId]: %w", admission.ErrInvalidRequest)
+		return field.Forbidden(fieldPath,
+			"binding must target either a user [userName]/[userPrincipalName] OR a group [groupName]/[groupPrincipalName]")
 	}
 
 	if newPRTB.ProjectName == "" {
-		return fmt.Errorf("binding must have field projectName set: %w", admission.ErrInvalidRequest)
+		return field.Required(fieldPath.Child("projectName"), "")
+	}
+
+	if newPRTB.RoleTemplateName == "" {
+		return field.Required(fieldPath.Child("roleTemplateName"), "")
 	}
 
 	roleTemplate, err := a.roleTemplateResolver.RoleTemplateCache().Get(newPRTB.RoleTemplateName)
 	if err != nil {
-		return fmt.Errorf("unknown reference roleTemplate '%s': %w", newPRTB.RoleTemplateName, err)
+		return err
 	}
 
 	if roleTemplate.Locked {
-		return fmt.Errorf("referenced role '%s' is locked and cannot be assigned: %w", roleTemplate.DisplayName, admission.ErrInvalidRequest)
+		return field.Forbidden(fieldPath.Child("roleTemplate"), "role template cannot be locked")
+	}
+
+	const projectContext = "project"
+	if roleTemplate.Context != projectContext {
+		return field.NotSupported(fieldPath.Child("roleTemplate", "context"), roleTemplate.Context, []string{projectContext})
```

Comment:
I am unsure if we can do this today with the existing rancher behavior.

Reply:
That's fine. Again, it's just transferring over the current behavior...

Reply:
Sorry if this is current behavior. We can leave it. Based on the PR comment, I thought this was new.